gstorm/include/detail/ranges/vector.h

```diff
@@ -45,182 +40,187 @@ struct range_forward_traits {
-      if (other.hasAccessor())
-        accessor.emplace(other.accessor.value());
-      else {
+  iterator(const iterator& other) : has_accessor(other.has_accessor), accessor(other.accessor), owner(other.owner), it(other.it), id_(){
+    if (!other.hasAccessor()) {
 #ifndef __SYCL_DEVICE_ONLY__
-        if (owner != 0)
-          _id = (reinterpret_cast<gvector*>(owner))->registerIterator(*this); // this should never be executed on the device
+      if (owner)
+        id_ = (reinterpret_cast<gvector<T>*>(owner))->registerIterator(*this); // this should never be executed on the device
 #endif
-      }
     }
-    //iterator(iterator&& other) : accessor(std::move(other.accessor)), it(other.it){ }
+  }
+  //iterator(iterator&& other) : accessor(std::move(other.accessor)), it(other.it){ }
 
-    bool hasAccessor() const { return static_cast<bool>(accessor); }
+  bool hasAccessor() const { return has_accessor; }
 
-    sycl_accessor_type getAccessor() const { return accessor.value(); }
+  sycl_accessor_type getAccessor() const { return accessor; }
 
-    void setAccessor(sycl_accessor_type acc) { accessor.emplace(acc); }
+  void setAccessor(sycl_accessor_type acc) { has_accessor = true; accessor = acc; }
 
-    explicit iterator(sycl_accessor_type acc, size_t offset = 0, ptrdiff_t ref_back = 0) : accessor(acc),
-                                                                                           owner(ref_back),
-                                                                                           it(offset),
-                                                                                           _id(-1) {}
+  explicit iterator(sycl_accessor_type acc, size_t offset = 0, ptrdiff_t ref_back = 0) :
+    has_accessor(true),
+    accessor(acc),
+    owner(ref_back),
+    it(offset),
+    id_(size_t_max) {}
 
-    reference operator*() {
-      return accessor.value()[it];
-    }
+  reference operator*() {
+    return accessor[it];
+  }
 
-    reference operator*() const {
-      return accessor.value()[it];
-    }
+  reference operator*() const {
+    return accessor[it];
+  }
 
-    iterator &operator++() {
-      ++it;
-      return *this;
-    }
+  iterator &operator++() {
+    ++it;
+    return *this;
+  }
 
-    iterator operator++(int) const {
-      auto ip = it;
-      ++it;
-      if (!hasAccessor())
-        return iterator(ip);
-      return iterator(accessor.value(), ip, owner);
-    }
+  iterator operator++(int) {
+    auto ip = it;
+    ++it;
+    return iterator(accessor, ip, owner);
+  }
 
-    iterator &operator--() {
-      --it;
-      return *this;
-    }
+  iterator &operator--() {
+    --it;
+    return *this;
+  }
 
-    iterator operator--(int) const {
-      auto ip = it;
-      --it;
-      if (!hasAccessor())
-        return iterator(ip);
-      return iterator(accessor.value(), ip, owner);
-    }
+  iterator operator--(int) {
+    auto ip = it;
+    --it;
+    return iterator(accessor, ip, owner);
+  }
```

Comment:
we need some documentation on how this iterator works (and specifically why is needed). This should go either on code comments or (better) on a separate readme file. We can use that later for the paper if we find the need to explain things in detail.

Reply:
@pacxx Could you help out with this?

Reply:
I am currently on vacation till January and have not much time between three kids. I can start when I am back at the office

Reply:
I am currently on vacation till January and have not much time between three kids. I can start when I am back at the office